### PR TITLE
18/E2eTestsWithDetox

### PR DESCRIPTION
**Pull request by alvrba**

### CODE REVIEW

**Comment by alvrba** [comment]
**The tests are also available in the EAS service (without running them locally):**

https://expo.dev/accounts/graasp/projects/graasp-mobile-builder/builds/352057ac-d340-416b-8dbb-46dd22eee39e
eas android detox

https://expo.dev/accounts/graasp/projects/graasp-mobile-builder/builds/a7bf8b7f-2269-43bd-b6ef-266da653601d
eas ios detox

**Comment by alvrba** [comment]
item tests
membership tests
global tests

File: .eslintrc [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
-}
\ No newline at end of file
+  "extends": ["universe", "prettier"],
+  "plugins": ["import", "jsx-a11y", "react"],
+  "env": {
+    "node": true,
+    "es2022": true
+  },
+  "rules": {
+    "import/order": "off",
```
Comment (pyphilia) [inline .eslintrc:9]: Remove.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: checkout
+        uses: actions/checkout@v4
+
+      - name: set up node
+        uses: actions/setup-node@v4
+        with:
+          node-version: 20
```
Comment (pyphilia) [inline .github/workflows/build.yml:16]: Remove.

File: app.config.js [hunk 3]
```diff
@@ -32,7 +39,7 @@ module.exports = {
     supportsTablet: true,
     bundleIdentifier: 'org.graasp.mobile',
     buildNumber: '2',
-    googleServicesFile: process.env.GOOGLESERVICE_INFO_PLIST,
+    googleServicesFile: process.env.EXPO_PUBLIC_GOOGLESERVICE_INFO_PLIST,
```
Comment (pyphilia) [inline app.config.js:42]: Update readme?

File: app.config.js [hunk 4]
```diff
@@ -48,7 +55,7 @@ module.exports = {
       backgroundColor: '#ffffff',
     },
     package: 'org.graasp.mobile',
-    googleServicesFile: process.env.GOOGLE_SERVICES_JSON,
+    googleServicesFile: process.env.EXPO_PUBLIC_GOOGLE_SERVICES_JSON,
```
Comment (pyphilia) [inline app.config.js:58]: Update Readme?

File: e2e/utils/openApp.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
+const appConfig = require('../../app.config.js');
+const { resolveConfig } = require('detox/internals');
+
+const platform = device.getPlatform();
+
+module.exports.openApp = async function openApp() {
```
Comment (pyphilia) [inline e2e/utils/openApp.ts:6]: What's this syntax? Does this mean `export default` ?

File: e2e/utils/openApp.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      url: deepLinkUrl,
+    });
+  } else {
+    await device.launchApp({
+      newInstance: true,
+      url: deepLinkUrl,
+    });
+  }
+
+  await sleep(3000);
```
Comment (pyphilia) [inline e2e/utils/openApp.ts:39]: Factor out.

File: e2e/utils/openApp.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+const getDeepLinkUrl = (url: string) =>
+  `exp+graasp-mobile-builder://expo-development-client/?url=${encodeURIComponent(url)}`;
+
+const getDevLauncherPackagerUrl = (platform: string) =>
+  `http://localhost:8081?platform=${platform}&dev=true&minify=false&disableOnboarding=1`;
```
Comment (pyphilia) [inline e2e/utils/openApp.ts:46]: Why `8081`? Can this be an env variable? Or a constant for test?
In fact I don't know what that link is used for.

File: e2e/utils/openApp.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+const getLatestUpdateUrl = () =>
+  `https://u.expo.dev/${getAppId()}?channel-name=test_debug&disableOnboarding=1`;
+
+const getAppId = () => appConfig?.expo?.extra?.eas?.projectId ?? '';
+
+export const sleep = (t: number) => new Promise(res => setTimeout(res, t));
```
Comment (pyphilia) [inline e2e/utils/openApp.ts:53]: Should this go in `utils.ts` instead?

File: eas.json [hunk 9]
```diff
@@ -3,15 +3,23 @@
     "version": ">= 3.1.0"
   },
   "build": {
+    "base": {
+      "node": "20.9.0"
```
Comment (pyphilia) [inline eas.json:7]: Can we set like "20.x" so that it uses the latest one?

File: eas.json [hunk 10]
```diff
@@ -20,6 +28,40 @@
...
+        "gradleCommand": ":app:assembleRelease :app:assembleAndroidTest -DtestBuildType=release",
+        "withoutCredentials": true
+      },
+      "ios": {
+        "simulator": true
+      },
+      "env": {
+        "EXPO_PUBLIC_DETOX_MOCKED": "mocked",
+        "EXPO_PUBLIC_AUTH_HOST": "https://auth.stage.graasp.org",
+        "EXPO_PUBLIC_API_HOST": "https://api.stage.graasp.org"
```
Comment (pyphilia) [inline eas.json:44]: I think this is equivalent as using an env file. Is there a way to use `.env.test`?

File: eas.json [hunk 11]
```diff
@@ -20,6 +28,40 @@
...
+      },
+      "ios": {
+        "buildConfiguration": "Debug",
+        "simulator": true
+      },
+      "env": {
+        "EXPO_USE_UPDATES": "1",
+        "EXPO_PUBLIC_DETOX_MOCKED": "mocked",
+        "EXPO_PUBLIC_AUTH_HOST": "https://auth.stage.graasp.org",
+        "EXPO_PUBLIC_API_HOST": "https://api.stage.graasp.org"
```
Comment (pyphilia) [inline eas.json:62]: Same as above.

File: metro.config.js [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+const config = getDefaultConfig(__dirname);
+
+const defaultSourceExts = require('metro-config/src/defaults/defaults').sourceExts;
+
+if (process.env.EXPO_PUBLIC_DETOX_MOCKED === 'mocked'){
+  config.resolver.sourceExts = ['mock.tsx', ...defaultSourceExts];
+}
+
+
+module.exports = config;
```
Comment (pyphilia) [inline metro.config.js:14]: I think I prefer this syntax, maybe let's not depend on a fixed string "mocked" but rather `true`? But wouldn't we want detox to be always mocked? Maybe not let me know. 
https://github.com/graasp/graasp-mobile-builder/blob/48f5eaba1cbf9d62e25f9e28ebdbea7177da1b4e/metro.config.js

File: package.json [hunk 13]
```diff
@@ -49,38 +57,53 @@
...
+    "prettier": "3.0.3",
+    "react-test-renderer": "18.2.0",
+    "ts-jest": "^29.1.1",
+    "typescript": "^5.2.2"
   },
-  "private": true
+  "private": true,
+  "engines": {
+    "node": ">=20"
+  }
```
Comment (pyphilia) [inline package.json:108]: Let's put that at the top of the file.

File: src/navigation/DrawerNavigator.tsx [hunk 14]
```diff
@@ -50,7 +53,15 @@ const CustomDrawerContent: FC<CustomDrawerContentProps> = (props: any) => {
         icon={({ color, size, focused }) => (
           <Feather name="external-link" size={size} color={color} />
         )}
-        onPress={() => Linking.openURL('https://graasp.eu/')}
+        onPress={() =>
+          Linking.openURL('mailto:contact@graasp.org').catch(() => {
```
Comment (pyphilia) [inline src/navigation/DrawerNavigator.tsx:57]: Factor out the email.

File: src/navigation/DrawerNavigator.tsx [hunk 15]
```diff
@@ -50,7 +53,15 @@ const CustomDrawerContent: FC<CustomDrawerContentProps> = (props: any) => {
...
+            Toast.show({
+              text1: 'Sorry, we could not open your mail app',
+              text2: 'Send an email to contact@graasp.org',
```
Comment (pyphilia) [inline src/navigation/DrawerNavigator.tsx:60]: And reuse here.

File: src/navigation/TabNavigator.tsx [hunk 16]
```diff
@@ -42,8 +45,14 @@ const TabNavigator = () => {
         options={{
           tabBarLabel: t('Shared Items')!,
           tabBarIcon: ({ color, size }) => (
-            <MaterialIcons name="folder-shared" size={size} color={color} />
+            <MaterialIcons
+              name="folder-shared"
+              size={size}
+              color={color}
+              testID="shared-items"
```
Comment (pyphilia) [inline src/navigation/TabNavigator.tsx:52]: Factor out?

File: src/screens/SharedScreen.tsx [hunk 17]
```diff
@@ -33,7 +40,11 @@ const SharedScreen: FC<SharedStackSharedProps> = ({ navigation }) => {
   }
 
   return (
-    <SafeAreaView style={styles.container} edges={['left']}>
+    <SafeAreaView
+      style={styles.container}
+      edges={['left']}
+      testID="shared-items-screen"
```
Comment (pyphilia) [inline src/screens/SharedScreen.tsx:46]: Factor out?

File: src/screens/SignInScreen.tsx [hunk 18]
```diff
@@ -108,6 +112,7 @@ const SignInScreen: FC<SignInProps> = ({ route: { params } }) => {
           title="Sign in"
           disabled={isLoading}
           onPress={_handlePressLoginButtonAsync}
+          testID="sign-in"
```
Comment (pyphilia) [inline src/screens/SignInScreen.tsx:115]: Factor out?

**Review by pyphilia — CHANGES_REQUESTED** [review]
Thanks for all these tests, it will help us a lot!
I've never seen a "close app" function, do you need it?
Also I think your PR contains lots of previous merged commit, maybe let's rebase next time...

File: e2e/utils/openApp.ts (hunk 7 above)
Comment (alvrba) [reply e2e/utils/openApp.ts:46]: It's the Expo CLI Metro Server default port. I believe the only way of changing it is by specifying it in the `yarn start -dev-client` command
https://docs.expo.dev/more/expo-cli/

File: eas.json (hunk 10 above)
Comment (alvrba) [reply eas.json:44]: FOR LOCAL TEST: I created an `.env.test` file, so every time we want to run the tests in local you should run the server with `NODE_ENV=test yarn start --dev-client`. So now we have a `.env.development` and `.env.test` files.

FOR EAS TESTS: These lines are only used in the EAS service, they are executed when you create a new build using EAS with `eas build --profile test`, so there are 3 options:
- Factor out these lines in another file
- Use the .env.test file but we need to run the command as `NODE_ENV=test eas build --profile test` to load the envs correctly from the test environment
- Leave it as it is now

Updated README.

File: metro.config.js (hunk 12 above)
Comment (alvrba) [reply metro.config.js:14]: I refactored it with your syntax, please take a look

File: .github/workflows/build.yml (hunk 2 above)
Comment (alvrba) [reply .github/workflows/build.yml]: Screenshot 2023-11-19 at 22 39 52
How can we solve this?

Comment (spaenleh) [reply .github/workflows/build.yml]: Why is it using yarn v1 ?
Update to latest version of yarn: `yarn set version latest` and then commit the files generated.

**Pull request merged**